README.md
# heror
![Build](https://badgen.net/travis/ozylog/heror/master)
![Coverage](https://badgen.net/coveralls/c/github/ozylog/heror/master)
![DevDependencies](https://badgen.net/david/dev/ozylog/heror)
![LatestVersion](https://badgen.net/npm/v/heror)
![License](https://badgen.net/npm/license/heror)

HTTP error classes

## Install
```
yarn add heror
```

## Class
```
new BadRequestError(message: string = 'Bad Request', data?: any)
```
- `message` - error message
- `data` - optional input that you can use for additional information regarding the error

## Function
```
createError(statusCode: number, error: string)
```
- `statusCode` - http error status code
- `error` - official http error type/message

## List of error classes
We provide some common http errors below. If your case is not covered by the errors. You could create new error using `createError` function. Please see [usage examples](#usage-examples).

| Status Code | Error Class Name         |
|-------------|--------------------------|
| 400         | BadRequestError          |
| 401         | UnauthorizedError        |
| 403         | ForbiddenError           |
| 404         | NotFoundError            |
| 405         | MethodNotAllowedError    |
| 406         | NotAcceptableError       |
| 408         | RequestTimeoutError      |
| 409         | ConflictError            |
| 412         | PreconditionFailedError  |
| 422         | UnprocessableEntityError |
| 500         | InternalServerError      |
| 501         | NotImplementedError      |
| 503         | ServiceUnavailableError  |
| 507         | InsufficientStorageError |
| 508         | NotExtendedError         |

## Usage Examples
```
import { BadRequestError, UnauthorizedError, createError } from 'heror';

const URITooLongError = createError(414, 'URI Too Long');

async function userController(req, res, next) {
  if (!req.auth) {
    throw new UnauthorizedError('Authentication is required', {
      expose: true,
      details: {}
    });
  }
  /*
    it will throw error instance
    {
      statusCode: 401,
      error: 'Unauthorized',
      message: 'Authentication is required',
      data: {
        expose: true,
        details: {}
      },
      ...nativeErrorProperties
    }
  /*

  if (!req.params.id) throw new BadRequestError();
  /*
    it will throw error instance
    {
      statusCode: 400,
      error: 'Bad Request',
      message: 'Bad Request',
      ...nativeErrorProperties
    }
  */

  if (req.uri.length > 100) {
    throw new URITooLongError('Max URI Length is 100 characters', {
      uriLength: req.uri.length
    });
  }
  /*
    it will throw error instance
    {
      statusCode: 414,
      error: 'URI Too Long',
      message: 'Max URI Length is 100 characters',
      data: {
        uriLength: 111
      },
      ...nativeErrorProperties
    }
  /*

}

```

## License
MIT


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [1.0.0] - 2019-07-27
### Added
- Added badges

## [0.1.0] - 2019-07-17
### Added
- Initial implementation
- Test


test/Heror.test.ts
import Heror from './../src/Heror';

describe('#Heror(statusCode, error, message, data)', () => {
  describe(`when statusCode=400, error='Bad Request'`, () => {
    const statusCode = 400;
    const error = 'Bad Request';
    let heror: Heror | undefined;

    beforeAll(() => {
      heror = new Heror(statusCode, error);
    });

    test('should have statusCode 400', () => {
      expect(heror!.statusCode).toBe(statusCode);
    });

    test('should have error Bad Request', () => {
      expect(heror!.error).toBe(error);
    });

    test('should have message Bad Request', () => {
      expect(heror!.message).toBe(error);
    });

    test('should not have data', () => {
      expect(heror!.data).toBe(undefined);
    });
  });

  describe(`when statusCode=401, error='Unauthorized', message='Invalid authentication', data={ token: '123random' }`, () => {
    const statusCode = 401;
    const error = 'Unauthorized';
    const message = 'Invalid authentication';
    const data = { token: '123random' }
    let heror: Heror | undefined;

    beforeAll(() => {
      heror = new Heror(statusCode, error, message, data);
    });

    test('should have statusCode 401', () => {
      expect(heror!.statusCode).toBe(statusCode);
    });

    test('should have error Unauthorized', () => {
      expect(heror!.error).toBe(error);
    });

    test('should have message Invalid authentication', () => {
      expect(heror!.message).toBe(message);
    });

    test('should have data', () => {
      expect(heror!.data).toEqual(data);
    });
  });
});


test/index.test.ts
import * as index from './../src/index';

describe('#index', () => {
  describe('when import Heror class', () => {
    test('should exist', () => {
      expect(index.Heror).toBeDefined();
      expect(index.Heror.name).toBe('Heror');
    });
  });

  [
    'BadRequestError',
    'UnauthorizedError',
    'ForbiddenError',
    'NotFoundError',
    'MethodNotAllowedError',
    'NotAcceptableError',
    'RequestTimeoutError',
    'ConflictError',
    'PreconditionFailedError',
    'UnprocessableEntityError',
    'InternalServerError',
    'NotImplementedError',
    'ServiceUnavailableError',
    'InsufficientStorageError',
    'NotExtendedError'
  ].forEach((className) => {
    describe(`when import ${className} class`, () => {
      test('should exist', () => {
        expect(index[className as keyof typeof index]).toBeDefined();
      });
    });
  });

  describe('#createError(statusCode, error)', () => {
    const createError = index.createError;

    describe('when statusCode=414, error=URI Too Long', () => {
      const statusCode = 414;
      const error = 'URI Too Long';
      let URITooLongError: index.HerorClass | undefined;

      beforeAll(() => {
        URITooLongError = createError(statusCode, error);
      });

      test('should create error class', () => {
        expect(URITooLongError!.name).toBe('bound Heror');
      });

      describe('when create instance', () => {
        let uriError: index.Heror | undefined;

        beforeAll(() => {
          uriError = new URITooLongError!();
        });

        test('should be instance of Heror and Error', () => {
          expect(uriError!).toBeInstanceOf(Error);
          expect(uriError!).toBeInstanceOf(index.Heror);
        });

        test('should have correct property', () => {
          expect(uriError!.statusCode).toBe(statusCode);
          expect(uriError!.error).toBe(error);
          expect(uriError!.message).toBe(error);
          expect(uriError!.data).toBeUndefined();
        });
      });

      describe('when create instance with message and data', () => {
        const message = 'Max URI Length is 100 characters';
        const data = { uriLength: 111 };
        let uriError: index.Heror | undefined;

        beforeAll(() => {
          uriError = new URITooLongError!(message, data);
        });

        test('should be instance of Heror and Error', () => {
          expect(uriError!).toBeInstanceOf(Error);
          expect(uriError!).toBeInstanceOf(index.Heror);
        });

        test('should have correct property', () => {
          expect(uriError!.statusCode).toBe(statusCode);
          expect(uriError!.error).toBe(error);
          expect(uriError!.message).toBe(message);
          expect(uriError!.data).toEqual(data);
        });
      });
    });
  });
});


src/index.ts
import Heror from './Heror';
export { default as Heror } from './Heror';

export type HerorClass = new (message?: string | undefined, data?: any) => Heror;

// 4xx Errors
export const BadRequestError = Heror.bind(this, 400, 'Bad Request');
export const UnauthorizedError = Heror.bind(this, 401, 'Unauthorized');
export const ForbiddenError = Heror.bind(this, 403, 'Forbidden');
export const NotFoundError = Heror.bind(this, 404, 'Not Found');
export const MethodNotAllowedError = Heror.bind(this, 405, 'Method Not Allowed');
export const NotAcceptableError = Heror.bind(this, 406, 'Not Acceptable');
export const RequestTimeoutError = Heror.bind(this, 408, 'RequestTimeoutError');
export const ConflictError = Heror.bind(this, 409, 'Conflict');
export const PreconditionFailedError = Heror.bind(this, 412, 'Precondition Failed');
export const UnprocessableEntityError = Heror.bind(this, 422, 'Unprocessable Entity');

// 5xx Errors
export const InternalServerError = Heror.bind(this, 500, 'Internal Server Error');
export const NotImplementedError = Heror.bind(this, 501, 'Not Implemented');
export const ServiceUnavailableError = Heror.bind(this, 503, 'Service Unavailable');
export const InsufficientStorageError = Heror.bind(this, 507, 'Insufficient Storage');
export const NotExtendedError = Heror.bind(this, 510, 'Not Extended');

export function createError(statusCode: number, error: string): HerorClass {
  return Heror.bind(this, statusCode, error);
}
